### PR TITLE
Cite 157 edit concept is broken

### PR DESCRIPTION
**Pull request by Charishma0249**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

1. When I try to edit a concept (all fields are filled) with an invalid url and click save, the edit page shows again but no error message. 
2. When I try to add/edit a concept type with an invalid URL and click save, it gets saved successfully without any check.
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java [hunk 1]
```diff
@@ -49,12 +52,29 @@ public String show(Model model, Principal principal) {
 
     @RequestMapping(value = "/auth/concepts/add", method = RequestMethod.POST)
     public String post(@Validated @ModelAttribute("conceptForm") CitationConceptForm form, BindingResult result,
-            Model model, Principal principal) {
+            Model model, Principal principal, RedirectAttributes redirectAttributes) {
 
-        if (result.hasErrors()) {
+        if(result.hasErrors())
+        {
+            ListIterator<ObjectError> listIterator = result.getAllErrors().listIterator();
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:59]: I think, a simple foreach loop would have worked as well

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java [hunk 2]
```diff
@@ -49,12 +52,29 @@ public String show(Model model, Principal principal) {
...
+            
+            while(listIterator.hasNext()) {
+                ObjectError error = listIterator.next();
+                
+                if(error.getDefaultMessage().equals("must be a valid URL")) {
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:64]: Other error messages would need to be displayed as well

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java [hunk 3]
```diff
@@ -52,10 +55,26 @@ public String show(Model model, @PathVariable("conceptId") String conceptId, Aut
     public String post(Model model, @PathVariable("conceptId") String conceptId, Authentication authentication,
             @Valid @ModelAttribute("form") CitationConceptForm form, BindingResult result,
             RedirectAttributes redirectAttributes) {
-        if (result.hasErrors()) {
+        
+        if(result.hasErrors()) {
+            
+            ListIterator<ObjectError> listIterator = result.getAllErrors().listIterator();
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java:61]: Same as previous comment

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java [hunk 4]
```diff
@@ -52,10 +55,26 @@ public String show(Model model, @PathVariable("conceptId") String conceptId, Aut
...
+            
+            while(listIterator.hasNext()) {
+                
+                if(listIterator.next().getDefaultMessage().equals("must be a valid URL")) {
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java:65]: Same as previous comment

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/AddConceptTypeController.java [hunk 5]
```diff
@@ -43,8 +47,22 @@ public String show(Model model) {
 
     @RequestMapping(value = "/auth/concepts/types/add", method = RequestMethod.POST)
     public String post(@Validated @ModelAttribute("conceptTypeForm") ConceptTypeForm form, BindingResult result,
-            Model model, Principal principal) {
+            Model model, Principal principal, RedirectAttributes redirectAttributes) {
         if (result.hasErrors()) {
+            ListIterator<ObjectError> listIterator = result.getAllErrors().listIterator();
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/AddConceptTypeController.java:52]: Same as other controllers

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java [hunk 6]
```diff
@@ -52,6 +55,20 @@ public String post(Model model, @PathVariable("conceptTypeId") String typeId, Au
             @Validated @ModelAttribute("form") ConceptTypeForm form, BindingResult result,
             RedirectAttributes redirectAttributes) {
         if (result.hasErrors()) {
+            ListIterator<ObjectError> listIterator = result.getAllErrors().listIterator();
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java:58]: Same as other controllers

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java (hunk 1 above)
Comment (Charishma0249) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:59]: Yes that would, also, it's less cumbersome. Will update it.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java (hunk 2 above)
Comment (Charishma0249) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:64]: Makes sense, will make it generic

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java [hunk 7]
```diff
@@ -1,17 +1,20 @@
 package edu.asu.diging.citesphere.web.user.concepts;
 
 import java.security.Principal;
+import java.util.ListIterator;
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:4]: If it is not used anymore, you can remove the unused imports

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java [hunk 8]
```diff
@@ -52,8 +55,16 @@ public String post(Model model, @PathVariable("conceptTypeId") String typeId, Au
             @Validated @ModelAttribute("form") ConceptTypeForm form, BindingResult result,
             RedirectAttributes redirectAttributes) {
         if (result.hasErrors()) {
+            
+            for(ObjectError error : result.getAllErrors()) {
+                    redirectAttributes.addFlashAttribute("show_alert", true);
```
Comment (MaulikLim) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java:60]: Address the codefactor flags for indentation

Comment (Charishma0249) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java]: Done

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java (hunk 7 above)
Comment (Charishma0249) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java]: Done

**Review by MaulikLim — APPROVED** [review]

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java [hunk 9]
```diff
@@ -49,12 +51,19 @@ public String show(Model model, Principal principal) {
...
     public String post(@Validated @ModelAttribute("conceptForm") CitationConceptForm form, BindingResult result,
-            Model model, Principal principal) {
+            Model model, Principal principal, RedirectAttributes redirectAttributes) {
 
-        if (result.hasErrors()) {
+        if(result.hasErrors()) {
+            for(ObjectError error : result.getAllErrors()) {
+                redirectAttributes.addFlashAttribute("show_alert", true);
+                redirectAttributes.addFlashAttribute("alert_msg", error.getDefaultMessage());
+                redirectAttributes.addFlashAttribute("alert_type", "danger");
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:60]: error messages should be shown in the form below the input field they belong to.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java [hunk 10]
```diff
@@ -49,12 +51,19 @@ public String show(Model model, Principal principal) {
...
+            }
+            
             model.addAttribute("conceptForm", form);
-            return "auth/concepts/add";
+            return "redirect:/auth/concepts/add";
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/AddConceptController.java:64]: here it's shouldn't redirect

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java [hunk 11]
```diff
@@ -52,10 +53,20 @@ public String show(Model model, @PathVariable("conceptId") String conceptId, Aut
...
-        if (result.hasErrors()) {
+        
+        if(result.hasErrors()) {
+            
+            for(ObjectError error : result.getAllErrors()) {
+                redirectAttributes.addFlashAttribute("show_alert", true);
+                redirectAttributes.addFlashAttribute("alert_msg", error.getDefaultMessage());
+                redirectAttributes.addFlashAttribute("alert_type", "danger");
+               
+            }
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java:64]: see above

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java [hunk 12]
```diff
@@ -52,10 +53,20 @@ public String show(Model model, @PathVariable("conceptId") String conceptId, Aut
...
+            
             model.addAttribute("form", form);
-            return "auth/concepts/edit";
+            return "redirect:/auth/concepts/{conceptId}/edit";
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/EditConceptController.java:67]: see above

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/AddConceptTypeController.java [hunk 13]
```diff
@@ -43,10 +46,17 @@ public String show(Model model) {
...
     public String post(@Validated @ModelAttribute("conceptTypeForm") ConceptTypeForm form, BindingResult result,
-            Model model, Principal principal) {
+            Model model, Principal principal, RedirectAttributes redirectAttributes) {
         if (result.hasErrors()) {
+            
+            for(ObjectError error : result.getAllErrors()) { 
+                redirectAttributes.addFlashAttribute("show_alert", true);
+                redirectAttributes.addFlashAttribute("alert_msg", error.getDefaultMessage());
+                redirectAttributes.addFlashAttribute("alert_type", "danger");
+            }
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/AddConceptTypeController.java:56]: see above

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/AddConceptTypeController.java [hunk 14]
```diff
@@ -43,10 +46,17 @@ public String show(Model model) {
...
+            
             model.addAttribute("conceptTypeForm", form);
-            return "auth/concepts/types/add";
+            return "redirect:/auth/concepts/types/add";
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/AddConceptTypeController.java:59]: see above

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java [hunk 15]
```diff
@@ -52,8 +52,16 @@ public String post(Model model, @PathVariable("conceptTypeId") String typeId, Au
             @Validated @ModelAttribute("form") ConceptTypeForm form, BindingResult result,
             RedirectAttributes redirectAttributes) {
         if (result.hasErrors()) {
+            
+            for(ObjectError error : result.getAllErrors()) {
+                redirectAttributes.addFlashAttribute("show_alert", true);
+                redirectAttributes.addFlashAttribute("alert_msg", error.getDefaultMessage());
+                redirectAttributes.addFlashAttribute("alert_type", "danger");
+                      
+            }
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java:61]: see above

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java [hunk 16]
```diff
@@ -52,8 +52,16 @@ public String post(Model model, @PathVariable("conceptTypeId") String typeId, Au
...
+            
             model.addAttribute("form", form);
-            return "auth/concepts/types/edit";
+            return "redirect:/auth/concepts/types/{conceptTypeId}/edit";
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/concepts/types/EditConceptTypeController.java:64]: see above

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Pull request closed without merging**